File: spark-job/src/main/scala/com/dqs/config/RuleConfig.scala
package com.dqs.config

import upickle.default._

case class CheckConfig(checkType: String, min: Option[Int] = None)
object CheckConfig {
  // uPickle 3.x encodes Option[T] as arrays by default; use a plain number/null codec instead
  implicit val optIntRw: ReadWriter[Option[Int]] = readwriter[ujson.Value].bimap[Option[Int]](
    { case Some(n) => ujson.Num(n.toDouble); case None => ujson.Null },
    {
      case ujson.Num(n) => Some(n.toInt)
      case ujson.Null   => None
      case other        => throw new IllegalArgumentException(s"Expected number or null for Int field, got: $other")
    }
  )
  implicit val rw: ReadWriter[CheckConfig] = macroRW
}

case class RuleConfig(dataset: String, checks: Seq[CheckConfig])
object RuleConfig {
  implicit val rw: ReadWriter[RuleConfig] = macroRW

  def fromJson(json: String): RuleConfig =
    try read[RuleConfig](json)
    catch { case e: Exception =>
      throw new IllegalArgumentException(s"Failed to parse rule config: ${e.getMessage}", e)
    }
}

File: spark-job/src/main/java/com/dqs/config/RuleConfig.java
package com.dqs.config;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;

@JsonIgnoreProperties(ignoreUnknown = true)
public class RuleConfig {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final String dataset;
    private final List<CheckConfig> checks;

    @JsonCreator
    public RuleConfig(
            @JsonProperty("dataset") String dataset,
            @JsonProperty("checks") List<CheckConfig> checks) {
        this.dataset = dataset;
        this.checks = checks;
    }

    /** Parses a JSON string into a RuleConfig; throws IllegalArgumentException on malformed input. */
    public static RuleConfig fromJson(String json) {
        try {
            return MAPPER.readValue(json, RuleConfig.class);
        } catch (Exception e) {
            throw new IllegalArgumentException("Failed to parse rule config: " + e.getMessage(), e);
        }
    }

    public String getDataset() { return dataset; }
    public List<CheckConfig> getChecks() { return checks; }
}
